Project SLA: refactoring and improvement, to make it more generic

katyukha opened the pull request:
What is done:
1. Added 'start_field_id' field to 'project.sla' model same as
   'control_field_id' field. This will make simpler integration with
   objects that have no 'create_date' field, or such field have other
   name. Also this will allow to control not only time from
   confirmation, but any cobination of date fields of object (model)
2. Added sql constraints to check if 'limit_qty' and 'warn_qty' entered
   correctly.
3. Partly fixed bug when control field is of type 'date' not 'datetime'
   Partly, because of, may be, incorect calculation of time when not
   exact time specified. but now it will not throw error.
   May be start using UoMs here, to be able calculate time diferences
   in hours or days, what can be choosed by user?
4. Bugfix with user_id. Now there no errors will be thrown on attempt
   to create SLA for model, which has no 'user_id' field
5. Changed code, that checks for `doc.state != 'cancelled'` to
   `doc.state not in ('cancelled', 'cancel')` because of
   'cancelled' state is not standard in other apps, in most apps
   'cancel' state used instead
6. Views refactoring. Especialy tree view added for 'project.sla'
   model, which shows model name and control field names in tree.

katyukha committed:
Project SLA: refactoring and improvement, to make it more generic

What is done:
 1. Added 'start_field_id' field to 'project.sla' model same as
    'control_field_id' field. This will make simpler integration with
    objects that have no 'create_date' field, or such field have other
    name. Also this will allow to control not only time from
    confirmation, but any cobination of date fields of object (model)
 2. Added sql constraints to check if 'limit_qty' and 'warn_qty' entered
    correctly.
 3. Partly fixed bug when control field is of type 'date' not 'datetime'
    Partly, because of, may be, incorect calculation of time when not
    exact time specified. but now it will not throw error.
    May be start using UoMs here, to be able calculate time diferences
    in hours or days, what can be choosed by user?
 4. Bugfix with user_id. Now there no errors will be thrown on attempt
    to create SLA for model, which has no 'user_id' field
 5. Changed code, that checks for ```doc.state != 'cancelled'``` to
    ```doc.state not in ('cancelled', 'cancel')``` because of
    'cancelled' state is not standard in other apps, in most apps
    'cancel' state used instead
 6. Views refactoring. Especialy tree view added for 'project.sla'
    model, which shows model name and control field names in tree.

dreispt commented:
Great! Good to see contributions here.
One question, why 7.0 and not 8.0? Do you have this deployed in v7 right now?
Also, do you mind fixing the flake8 and pylint errors related to  project_sla` detected by the TravisCI build?

katyukha commented:
Thanks!
7.0 because of we still use 7.0 in production, and 8.0 in plans yet... there are a lot of our code to be ported.
This patch had been deployed today on production, so in next few days, may be some other fixes appear. If all will work fine, I may try to add this changes to 8.0 branch.
As for flake8 and pylint, seems, that #45 pull request do it.

dreispt commented:
#45 is now merged. Can you rebase?

katyukha commented:
Rebased!

katyukha commented:
Posibly, I will implement integration with stock.picking and stock.move objects on version 7.0
Imlement it as separate module in this repo? And It seems to be good to create new pull request for them.

dreispt commented:
@katyukha A new module, definitely.

katyukha committed:
Project SLA: Bugfixes

1. Refactored analytic_account.reapply_sla method
   added check for 'cancel' state. and removed unneccessary search for
   project_ids
2. Bugfix for calculationg of SLA status ported from 8.0 branch
3. In SLA Controlled model mixin, store SLA logic moved to separate
   method

katyukha committed:
Project SLA: Added menu 'Project / Config / SLA' + bugfix

Fixed bug: context recived by method was modified inside, thus multiple call to
method ```store_sla_control``` of object 'project.sla.control' had no
sense.

dreispt commented on project_sla/project_sla_control.py:
```diff
@@ -176,13 +177,31 @@ def _get_computed_slas(self, cr, uid, doc, context=None):
+            if field == 'control':
+                sla_field = sla.control_field_id
+            elif field == 'start':
+                sla_field = sla.start_field_id
```
or maybe just `sla_field = get_attr(sla, field)`  :smile_cat:

katyukha replied:
Do You mean `sla_field = getattr(sla, '%s_field_id' % field)`?

dreispt replied:
I guess I misread, forget it.

dreispt commented:
LGTM, but could you fix PEP8 issues?:

```
./project_sla/project_sla_control.py:147:80: E501 line too long (113 > 79 characters)
./project_sla/project_sla_control.py:148:80: E501 line too long (104 > 79 characters)
./project_sla/project_sla_control.py:190:80: E501 line too long (89 > 79 characters)
./project_sla/project_sla_control.py:333:46: E231 missing whitespace after ','
./project_sla/project_sla_control.py:333:51: E231 missing whitespace after ','
./project_sla/project_sla_control.py:334:80: E501 line too long (85 > 79 characters)
./project_sla/project_sla_control.py:336:80: E501 line too long (99 > 79 characters)
```

katyukha commented:
I hope that in monday-tuesday I will do changes.
And also there are already implemented report view for SLA items and basic stock picking/move integration. So I plan to create two more pull requests for that changes, one for SLA Report view and one for stock integration. Is it right way?

dreispt commented:
@katyukha Yes, it's best to try keeping PRs smaller, or reviews will be hard.

katyukha commented:
Ok)

katyukha committed:
Project SLA: PEP8 fixes + .travis.yml fixes

katyukha commented:
PEP8 seems to be fixed for project_sla module.

dreispt commented:
We should fix #55 to make TravisCI builds work again before doing any merge in branch 7.0.

katyukha commented:
Rebased on current 7.0

katyukha commented:
There are changes from #58 here. Build have ran unit tests two times. Leave unit tests for odoo/odoo repo only, or keep unittests for OCA/OCB too? And there strange error in unit tests for OCA/OCB, but it seems that some packages (psycopg) was not installed.

dreispt commented on .travis.yml:
```diff
@@ -6,9 +6,9 @@ virtualenv:
-  - VERSION="7.0" ODOO_REPO="odoo/odoo" UNIT_TEST="1"
+  - VERSION="7.0" LINT_CHECK="1"
+  - VERSION="7.0" ODOO_REPO="odoo/odoo" LINT_CHECK="0" UNIT_TEST="1"
+  - VERSION="7.0" ODOO_REPO="OCA/OCB" LINT_CHECK="0" UNIT_TEST="1"
```
Integrated tests are missing. This should be:

```
- VERSION="7.0" LINT_CHECK="1"
- VERSION="7.0" ODOO_REPO="odoo/odoo" LINT_CHECK="0"
- VERSION="7.0" ODOO_REPO="OCA/OCB" LINT_CHECK="0"
- VERSION="7.0" UNIT_TEST="1" LINT_CHECK="0"
```

dreispt commented:
I restarted the build and it's green now.
It would be nice if you addressed my comment on `.travis.yml`, but all :+1: otherwise.

katyukha commented:
As for `.travis.yml` i add changes. But do not kanow what to deal with that error(

dreispt commented:
Great! The error is on the `project_categ` and is unrelated.
No problem merging this.
Again :+1:

dreispt commented:
I believe no one else will adventure in this review.
This is open for a long time, and blocks further work on SLAs.
If no one else opposes, I'll merge it.
